### ISSUE TITLE
ALTER TABLE allows changing datatype of distribution key to an unsupported one

### ISSUE THREAD

**Issue by hlinnaka** [greenplum-db/gpdb#6317]
```
postgres=# CREATE TABLE distpol_typechange (p text) DISTRIBUTED BY (p);
CREATE TABLE
postgres=# ALTER TABLE distpol_typechange ALTER COLUMN p TYPE point USING p::point;
ALTER TABLE
postgres=# insert into distpol_typechange values ('(1,1)');
ERROR:  cannot use expression as distribution key, because it is not hashable (cdbmutate.c:1329)
```

**Comment by hlinnaka** [comment]
A similar case:
```
postgres=# create table tstab (i int4, t tsvector) distributed by (i);
CREATE TABLE
postgres=# create unique index tstab_idx on tstab(t);
NOTICE:  updating distribution policy to match new unique index
CREATE INDEX
postgres=# insert into tstab values (1, 'foo');
ERROR:  cannot use expression as distribution key, because it is not hashable (cdbmutate.c:1329)
```

**Comment by hlinnaka** [comment]
This will be fixed by PR #6327. I'll just add a test case to that.

**Comment by hlinnaka** [comment]
Fixed by the big hash opclasses PR.
